### PR TITLE
Check for invalid index to avoid IOOBE

### PR DESCRIPTION
**Pull request by yschimke**
Avoid any externally noticeable RuntimeExceptions due to bad server data

### CODE REVIEW

**Comment by yschimke** [comment]
Addresses https://github.com/square/okhttp/issues/3793

**Comment by yschimke** [comment]
Main discussion points

- Is this worth a bigger refactor e.g. check this whenever accessing the dynamic table
- Should it check the null range (start of dynamic table) also?

File: okhttp/src/main/java/okhttp3/internal/http2/Hpack.java [hunk 1]
```diff
@@ -258,11 +258,16 @@ private void readLiteralHeaderWithIncrementalIndexingNewName() throws IOExceptio
...
     }
 
-    private ByteString getName(int index) {
+    private ByteString getName(int index) throws IOException {
       if (isStaticHeader(index)) {
         return STATIC_HEADER_TABLE[index].name;
       } else {
-        return dynamicTable[dynamicTableIndex(index - STATIC_HEADER_TABLE.length)].name;
+        int dynamicTableIndex = dynamicTableIndex(index - STATIC_HEADER_TABLE.length);
+        if (dynamicTableIndex < 0 || dynamicTableIndex > dynamicTable.length - 1) {
```
Comment (swankjesse) [inline okhttp/src/main/java/okhttp3/internal/http2/Hpack.java]: nit: more conventional to use `>=` instead of `>` and `- 1` ?

**Comment by swankjesse** [comment]
We probably want to fuzz the decoder with near-real data?

**Comment by yschimke** [comment]
Do you think it's better to fuzz the decoder and various components, in which case potentially finding theoretical but maybe inaccessible points caught by earlier components? Or to use something like a FuzzingSocket and test end to end?

**Pull request merged**